File: cs/2_3_adapter.cs
using System;

namespace Structural.Adapter
{
    public interface ITarget
    {
        string GetRequest(string Request);
    }

    class Adaptee
    {
        public string GetSpecificRequest(string Request, int RequestLength, bool MustBeTrue)
        {
            if (!MustBeTrue)
                throw new Exception("MustBeTrue is not true");
            if (Request.Length != RequestLength)
                throw new Exception("RequestLength is wrong");
            return $"Responce for {Request}";
        }
    }

    class Adapter : ITarget
    {
        private readonly Adaptee _adaptee;
        public Adapter(Adaptee adaptee)
        {
            this._adaptee = adaptee;
        }

        public string GetRequest(string Request)
        {
            return $"This is adapted '{this._adaptee.GetSpecificRequest(Request, Request.Length, true)}'";
        }
    }
}

File: js/2_3_adapter.js
class Adaptee {
    getSpecificRequest(request, requestLength, mustBeTrue) {
        if (!mustBeTrue)
            throw "mustBeTrue is not true";
        if (request.length != requestLength)
            throw "requestLength is wrong";
        return `Responce for ${request}`;
    }
}

class Adapter
{
    constructor (adaptee)
    {
        if (adaptee && adaptee.getSpecificRequest)
            this._adaptee = adaptee;
        else 
            throw `Can't adapt ${adaptee}`;
    }

    getRequest(request)
    {
        return `This is adapted '${this._adaptee.getSpecificRequest(request, request.length, true)}'`;
    }
}

function adapt(adaptee){
    function adaptRequest(method){
        return function(request){
           return `This is adapted '${method.apply(adaptee, [request, request.length, true])}'`;
        }
    }

    if (!(adaptee && adaptee.getSpecificRequest))
        throw `can't adapt ${adaptee}`;
    adaptee.getRequest = adaptRequest(adaptee.getSpecificRequest);
    return adaptee;
}

export {Adaptee, Adapter, adapt};